### PR TITLE
Add LayerNorm support for Vivado

### PR DESCRIPTION
**Pull request by rianbrooksflynn**
# Description

This PR adds support for Layer Normalization using either Keras or PyTorch with the Vivado backend in `io_parallel` mode.

This implementation uses a lookup table for inverse square root; the inputs to the lookup table follow a logarithmic distribution for better accuracy.

Tests have been added for both Keras and Pytorch parsing.

Credit is due to @Ethan0Jiang and @LostEcho365 (Zhixing Jiang and Dennis Yin) for their Vivado implementation and Keras parsing support; my contributions were making a change to the inverse square root lookup table implementation, implementing PyTorch support, and adding unit tests. (Here's a [link](https://arxiv.org/abs/2409.05207) to their pre-print.) The original code authors have given permission for their code to be merged into hls4ml.

Linked issue: https://github.com/fastmachinelearning/hls4ml/issues/1109

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] A new research paper code implementation

## Tests

Two unit tests added: `test/pytest/test_layernorm.py` and `test/pytest/test_layernorm_pytorch.py`

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: hls4ml/converters/keras/core.py
Comment (rianbrooksflynn) [inline hls4ml/converters/keras/core.py:1]: I'm not sure why the diff is so broken, maybe something to do with the line endings?

File: hls4ml/model/profiling.py
Comment (rianbrooksflynn) [inline hls4ml/model/profiling.py:1]: Diff also broken here, maybe line endings issue

**Comment by JanFSchulte** [comment]
The pytest failure is the QKeras tests timing out. There's 299 tests being run in that batch, which I guess is too many. Is there a why to reshuffle the batches to avoid the timeout?

**Comment by rianbrooksflynn** [comment]
pre-commit.ci autofix

**Comment by jmitrevs** [comment]
For the "broken" diffs, we should look to see what the whitespace error is and fix it before merging.

**Comment by JanFSchulte** [comment]
I did a bit of digging and it's not a whitespace problem but rather the file endings are improperly encoded. Likely the person we got the code from was using a Windows machine. @rianbrooksflynn, you can install the  `dos2unix` package and just do `dos2unix file.name` to fix this.

File: test/pytest/test_layernorm.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    model.compile()
+    return model
+
+
+# Currently only Vivado in io_parallel mode is supported
+def test_layernorm(model, data):
+    config = hls4ml.utils.config_from_keras_model(model, granularity='name', backend='Vivado')
+    output_dir = str(test_root_path / 'hls4mlprj_layernorm_Vivado_io_parallel')
+    hls_model = hls4ml.converters.convert_from_keras_model(
+        model, backend='Vivado', hls_config=config, io_type='io_parallel', output_dir=output_dir
```
Comment (jmitrevs) [inline test/pytest/test_layernorm.py:35]: If we are testing one backend only from (Vivado, Vitis), I would prefer that the backend be Vitis, since that will be the future focus more.

**Review by jmitrevs — APPROVED** [review]

**Comment by jmitrevs** [comment]
We should squash the commits when we merge this.

File: hls4ml/backends/vivado/passes/core_templates.py [hunk 2]
```diff
@@ -119,6 +128,59 @@ def format(self, node):
         return self.template.format(**params)
 
 
+# LayerNormalization templates
+
+layernorm_config_template = """struct config{index} : nnet::layernorm_config {{
+    static const unsigned n_in = {n_in};
+    static const unsigned seq_len = {seq_len};
+    static const unsigned table_size = {table_size};
+    static constexpr double table_range = {table_range};
```
Comment (vloncar) [inline hls4ml/backends/vivado/passes/core_templates.py:137]: Any reason this is a double? It is not used as such, and breaks Vivado synthesis.

File: hls4ml/backends/vivado/passes/core_templates.py [hunk 3]
```diff
@@ -119,6 +128,59 @@ def format(self, node):
...
+    static const unsigned io_type = nnet::{iotype};
+    static const unsigned reuse_factor = {reuse};
+    static const bool store_weights_in_bram = false;
+    static constexpr double epsilon = {epsilon};
```
Comment (vloncar) [inline hls4ml/backends/vivado/passes/core_templates.py:141]: Similar to above, this will only ever be used to store values in `table_t`, so we should ensure epsilon is compatible

File: hls4ml/backends/vivado/vivado_backend.py [hunk 4]
```diff
@@ -558,6 +559,21 @@ def init_softmax(self, layer):
...
             ), 'Softmax with io_parallel strategy cannot be used on multidimensional tensors.'
 
+    @layer_optimizer(LayerNormalization)
+    def init_layernormalization(self, layer):
+        if 'table_t' not in layer.attributes:
+            layer.set_attr(
+                'table_t', NamedType(name=layer.name + '_table_t', precision=FixedPrecisionType(width=16, integer=6))
+            )
+        if 'table_size' not in layer.attributes:
+            layer.set_attr('table_size', 4096)  # table size
```
Comment (vloncar) [inline hls4ml/backends/vivado/vivado_backend.py:569]: These attributes should be set as default in `_register_layer_attributes()` not here.

Also, is 4096 necessary for this implementation to work? All other tables are 1024.

File: hls4ml/converters/keras/core.py [hunk 5]
```diff
@@ -129,6 +129,34 @@ def parse_batchnorm_layer(keras_layer, input_names, input_shapes, data_reader):
...
+        raise Exception('assigning the axis is not currently supported by hls4ml, only axis 2 is supported')
+
+    layer['gamma_data'] = get_weights_data(data_reader, layer['name'], 'gamma')
+    layer['beta_data'] = get_weights_data(data_reader, layer['name'], 'beta')
+
+    layer['epsilon'] = keras_layer['config']['epsilon']
+    if layer['epsilon'] <= 0:
+        raise Exception('epsilon must be positive')
+
+    return layer, [shape for shape in input_shapes[0]]
```
Comment (vloncar) [inline hls4ml/converters/keras/core.py:157]: This function should also parse `axis` parameter to avoid misparsing. Sure, we only support `axis=-1` and we can raise exceptions about it, but should be handled.

File: hls4ml/converters/pytorch/core.py [hunk 6]
```diff
@@ -138,3 +138,32 @@ def parse_batchnorm_layer(operation, layer_name, input_names, input_shapes, node
...
+    layer['name'] = layer_name
+    layer['inputs'] = input_names
+
+    in_size = 1
+    for dim in input_shapes[0][1:]:
+        in_size *= dim
+    layer['n_in'] = layer['n_out'] = in_size
+
+    if not ((len(input_shapes[0])) == 3):
+        raise Exception('input size is not currently supported by hls4ml, only dim3 is supported')
```
Comment (vloncar) [inline hls4ml/converters/pytorch/core.py:159]: What is `dim3`? This sentence could be made a bit nicer :-)

File: hls4ml/model/layers.py [hunk 7]
```diff
@@ -1058,6 +1058,30 @@ def add_bias(self, bias, quantizer=None, precision=None):
         self.add_weights_variable(name='bias', var_name='b{index}', data=bias, quantizer=quantizer, precision=precision)
 
 
+class LayerNormalization(Layer):
+    _expected_attributes = [
+        Attribute('n_in'),
+        Attribute('seq_len'),
+        Attribute('epsilon', value_type=float, default=1e-3),
```
Comment (vloncar) [inline hls4ml/model/layers.py:1065]: `axis` too

File: hls4ml/templates/vivado/nnet_utils/nnet_layernorm.h [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    typedef float scale_t;
+
+    // Layer Sizes
+    static const unsigned n_in = 20;
+    static const unsigned seq_len = 4;
+
+    // Resource reuse info
+    static const unsigned io_type = io_parallel;
+    static const unsigned reuse_factor = 1;
+    static const bool store_weights_in_bram = false;
```
Comment (vloncar) [inline hls4ml/templates/vivado/nnet_utils/nnet_layernorm.h:25]: Can we tidy up this example config so that it has the fields that are actually expected (like all types and ranges) and not the ones that are not used (like weights_in_bram)

**Review by vloncar — CHANGES_REQUESTED** [review]
Looks good, could use some minor cosmetics.

File: test/pytest/test_layernorm.py (hunk 1 above)
Comment (rianbrooksflynn) [reply test/pytest/test_layernorm.py]: Testing both backends now :)

File: hls4ml/backends/vivado/passes/core_templates.py (hunk 2 above)
Comment (rianbrooksflynn) [reply hls4ml/backends/vivado/passes/core_templates.py]: I've changed it to an `unsigned` type.

File: hls4ml/backends/vivado/passes/core_templates.py (hunk 3 above)
Comment (rianbrooksflynn) [reply hls4ml/backends/vivado/passes/core_templates.py]: Changed to `unsigned` type.

File: hls4ml/backends/vivado/vivado_backend.py (hunk 4 above)
Comment (rianbrooksflynn) [reply hls4ml/backends/vivado/vivado_backend.py]: Unfortunately according to my tests, 4096 is necessary to achieve absolute differences of < 0.05 in accuracy.

File: hls4ml/converters/keras/core.py (hunk 5 above)
Comment (rianbrooksflynn) [reply hls4ml/converters/keras/core.py:162]: Sounds good, done.

File: hls4ml/converters/pytorch/core.py (hunk 6 above)
Comment (rianbrooksflynn) [reply hls4ml/converters/pytorch/core.py]: Changed the wording :)

File: hls4ml/model/layers.py (hunk 7 above)
Comment (rianbrooksflynn) [reply hls4ml/model/layers.py]: Done

File: hls4ml/templates/vivado/nnet_utils/nnet_layernorm.h (hunk 8 above)
Comment (rianbrooksflynn) [reply hls4ml/templates/vivado/nnet_utils/nnet_layernorm.h]: Done

File: test/pytest/test_layernorm.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+@pytest.fixture(scope='module')
+def data():
+    np.random.seed(0)
+    return np.random.rand(100, *in_shape)
+
+
+@pytest.fixture(scope='module')
+def model():
+    model = Sequential()
+    model.add(LayerNormalization(input_shape=in_shape))
```
[the remark this answers is not in the document]
Comment (rianbrooksflynn) [reply test/pytest/test_layernorm.py]: Made a new test with changed config

File: test/pytest/test_layernorm.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    model.compile()
+    return model
+
+
+# Currently only Vivado/Vitis in io_parallel mode is supported
+@pytest.mark.parametrize('backend', ['Vivado', 'Vitis'])
+def test_layernorm(model, data, backend):
+    config = hls4ml.utils.config_from_keras_model(model, granularity='name', backend=backend)
```
[the remark this answers is not in the document]
Comment (rianbrooksflynn) [reply test/pytest/test_layernorm.py:70]: Made a new test with changed config

File: test/pytest/test_layernorm_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+@pytest.fixture(scope='module')
+def data():
+    np.random.seed(0)
+    return np.random.rand(100, *in_shape)
+
+
+@pytest.fixture(scope='module')
+def model():
+    model = nn.Sequential(nn.LayerNorm(in_shape[-1]))
```
[the remark this answers is not in the document]
Comment (rianbrooksflynn) [reply test/pytest/test_layernorm_pytorch.py]: Made a new test with changed config

**Comment by rianbrooksflynn** [comment]
Thanks @vloncar for your review and apologies that it took me until now to get around to implementing your feedback! I would appreciate a second look from you at this pull request.

**Comment by JanFSchulte** [comment]
Fixed the latest merge conflicts. @vloncar, can you check if Rian's fixes are enough to have this merged?

**Comment by The-Padi** [comment]
Any update on the merge ?

**Comment by JanFSchulte** [comment]
pre-commit.ci autofix

**Comment by JanFSchulte** [comment]
I did not observe any warning about the scheduling. The timing results look all very similar:
image

File: test/pytest/test_layernorm.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    custom_config = hls4ml.utils.config_from_keras_model(custom_epsilon_model, granularity='name', backend=backend)
+    custom_config['LayerName']['layer_normalization']['Precision']['accum'] = 'ap_fixed<10,4>'
+    custom_config['LayerName']['layer_normalization']['table_t'] = 'ap_fixed<12,5>'
+    custom_config['LayerName']['layer_normalization']['TableSize'] = 2048
+    custom_config['LayerName']['layer_normalization']['TableRangePower2'] = 1
+    output_dir = str(test_root_path / f'hls4mlprj_layernorm_config_{backend}_io_parallel')
+    hls_model = hls4ml.converters.convert_from_keras_model(
+        custom_epsilon_model, backend=backend, hls_config=custom_config, io_type='io_parallel', output_dir=output_dir
+    )
+    hls_model.compile()
```
Comment (vloncar) [inline test/pytest/test_layernorm.py:56]: This test would be faster if we used `hls_model.write()`, or completely skip the step of writing/linking. We don't use it here. The later accuracy test checks if the produced code is compilable

File: hls4ml/converters/pytorch/core.py [hunk 13]
```diff
@@ -160,6 +162,42 @@ def parse_batchnorm_layer(operation, layer_name, input_names, input_shapes, node
...
+    layer['inputs'] = input_names
+
+    in_size = 1
+    for dim in input_shapes[0][1:]:
+        in_size *= dim
+    layer['n_in'] = layer['n_out'] = in_size
+
+    if not ((len(input_shapes[0])) == 3):
+        raise Exception(
+            'input size is not currently supported by hls4ml; '
```
Comment (vloncar) [inline hls4ml/converters/pytorch/core.py:182]: Would be good to say `Input shape <some shape> is not supported, only ...`
